### PR TITLE
Fix 7190 home work parking draft

### PR DESCRIPTION
**Pull request by Dima-1**

### CODE REVIEW

File: OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java [hunk 1]
```diff
@@ -682,6 +842,45 @@ protected void createDefaultCategories() {
 		addEmptyCategory(context.getString(R.string.shared_string_others));
 	}
 
+	public FavoriteGroup createPersonalGroup() {
```
Comment (max-klaus) [inline OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java]: Is method get called from anywhere?

File: OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java [hunk 2]
```diff
@@ -104,6 +111,96 @@ public void run() {
...
+	PointDescription getWorkPointDescription() {
+		return getWorkPoint().getPointStringDescription();
+	}
+
+	PointDescription getParkingPointDescription() {
+		return getParkingPoint().getPointStringDescription();
+	}
+
+
+	private FavouritePoint getPersonalPoint(String name) {
```
Comment (max-klaus) [inline OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java]: I suggest to store cachedHomePoint, cachedWorkPoint and cachedParkingPoint and not iterate all favorites.

File: OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java [hunk 3]
```diff
@@ -598,6 +752,12 @@ public int compare(FavouritePoint o1, FavouritePoint o2) {
 				if (res == 0) {
 					res = collator.compare(s1, s2);
 				}
+				if (s1.equals(WORK) && s2.equals(PARKING) && o1.getCategory().equals(PERSONAL) && o2.getCategory().equals(PERSONAL)) {
```
Comment (max-klaus) [inline OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java]: Something wrong with sorting. Home absent and it is not translation dependent.

File: OsmAnd/src/net/osmand/plus/AppInitializer.java [hunk 4]
```diff
@@ -194,6 +194,10 @@ public void initVariables() {
 				app.getSettings().migrateGlobalPrefsToProfile();
 				startPrefs.edit().putInt(VERSION_INSTALLED_NUMBER, VERSION_3_5).commit();
 			}
+			if (prevAppVersion < VERSION_3_5 || Version.getAppVersion(app).equals("3.5.3")) {
```
Comment (max-klaus) [inline OsmAnd/src/net/osmand/plus/AppInitializer.java]: Why duplicated if?

File: OsmAnd/src/net/osmand/plus/AppInitializer.java [hunk 5]
```diff
@@ -194,6 +194,10 @@ public void initVariables() {
...
+//				app.getSettings().migrateHomeWorkParkingToFavorites();
```
Comment (max-klaus) [inline OsmAnd/src/net/osmand/plus/AppInitializer.java]: Why commented?

File: OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java [hunk 6]
```diff
@@ -199,6 +296,43 @@ public boolean deleteFavourite(FavouritePoint p, boolean saveImmediately) {
...
+		if (hasParkingPoint()) {
+			getParkingPoint().setDescription(description.getName());
+			editFavourite(getParkingPoint(),latLon.getLatitude(),latLon.getLongitude());
+		} else {
+			addFavoritePersonal(latLon, PARKING, description.getName());
+		}
+	}
+
+
+	void addFavoritePersonal(LatLon latLon, String name, String description) {
```
Comment (max-klaus) [inline OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java]: Use annotations in new methods. Like @Nullable or @NonNull.

File: OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java [hunk 7]
```diff
@@ -199,6 +296,43 @@ public boolean deleteFavourite(FavouritePoint p, boolean saveImmediately) {
...
+		if (latLon != null) {
```
Comment (max-klaus) [inline OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java]: Bad approach. Why we even try to pass null to method addFavourite?

File: OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java [hunk 8]
```diff
@@ -104,6 +111,96 @@ public void run() {
...
+	public FavouritePoint getParkingPoint() {
+		return getPersonalPoint(PARKING);
+	}
+
+	LatLon getWorkPointLatLon() {
+		LatLon workPointLatLon;
+		if (hasWorkPoint()) {
+			workPointLatLon = new LatLon(getWorkPoint().getLatitude(), getWorkPoint().getLongitude());
+		} else {
+			workPointLatLon = new LatLon(0, 0);
```
Comment (max-klaus) [inline OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java]: If there is no point - return null

File: OsmAnd/src/net/osmand/plus/TargetPointsHelper.java [hunk 9]
```diff
@@ -33,13 +33,16 @@
 	private OsmandApplication ctx;
 	private TargetPoint homePoint = null;
```
Comment (max-klaus) [inline OsmAnd/src/net/osmand/plus/TargetPointsHelper.java]: Why do we need these points still? They must be deleted!

File: OsmAnd/src/net/osmand/plus/activities/FavoritesTreeFragment.java [hunk 10]
```diff
@@ -943,6 +974,27 @@ public void onClick(View v) {
 			name.setTypeface(Typeface.DEFAULT, visible ? Typeface.NORMAL : Typeface.ITALIC);
 			name.setTextColor(getResources().getColor(visible ? enabledColor : disabledColor));
 			distanceText.setText(distance);
+			if (groupPosition == 0) {
+				String address = model.getDescription() != null ? "• " + model.getDescription() : "";
+				distanceText.setText(distance + address);
```
Comment (max-klaus) [inline OsmAnd/src/net/osmand/plus/activities/FavoritesTreeFragment.java]: Use translations and placeholders.

File: OsmAnd/src/net/osmand/plus/activities/FavoritesTreeFragment.java [hunk 11]
```diff
@@ -943,6 +974,27 @@ public void onClick(View v) {
...
+				switch (model.getName()) {
+					case FavouritesDbHelper.HOME:
+						icon.setImageDrawable(getMyApplication().getUIUtilities().getThemedIcon(R.drawable.ic_action_home_dark));
```
Comment (max-klaus) [inline OsmAnd/src/net/osmand/plus/activities/FavoritesTreeFragment.java]: We have favorites in other parts of the app. Will it be copy-paste? BTW - looks like other places are not handled at all. For example - Search.

File: OsmAnd/src/net/osmand/plus/routepreparationmenu/cards/HomeWorkCard.java [hunk 12]
```diff
@@ -3,11 +3,14 @@
 import android.view.View;
 import android.widget.TextView;
 
+import net.osmand.data.FavouritePoint;
```
Comment (max-klaus) [inline OsmAnd/src/net/osmand/plus/routepreparationmenu/cards/HomeWorkCard.java]: Why these new imports here?

File: OsmAnd/src/net/osmand/plus/OsmandSettings.java [hunk 13]
```diff
@@ -252,6 +254,18 @@ public void migrateGlobalPrefsToProfile() {
 		}
 	}
 
+	public void migrateHomeWorkParkingToFavorites() {
+		FavouritesDbHelper favorites = ctx.getFavorites();
+		favorites.addFavoritePersonal(getHomePoint(),FavouritesDbHelper.HOME,getHomePointDescription().getName());
+		favorites.addFavoritePersonal(getWorkPoint(),FavouritesDbHelper.WORK,getWorkPointDescription().getName());
+		if (OsmandPlugin.getEnabledPlugin(ParkingPositionPlugin.class) != null) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/OsmandSettings.java]: Should not be here, no plugin class in settings. Either parking plugin should migrate by itself or the point could be lost after upgrade

File: OsmAnd/src/net/osmand/plus/TargetPointsHelper.java [hunk 14]
```diff
@@ -196,6 +203,35 @@ private void readHomeWorkFromSettings() {
...
+		if (favorites.hasWorkPoint()) {
+			workPoint = TargetPoint.create(favorites.getWorkPointLatLon(), new PointDescription(PointDescription.POINT_TYPE_LOCATION, PointDescription.getSearchAddressStr(ctx)));
+		}
+		if (favorites.hasParkingPoint()) {
+			parkingPoint = TargetPoint.create(favorites.getParkingPointLatLon(), new PointDescription(PointDescription.POINT_TYPE_LOCATION, PointDescription.getSearchAddressStr(ctx)));
+		}
+		if (!ctx.isApplicationInitializing()) {
+			lookupAddressForHomePoint();
+			lookupAddressForWorkPoint();
+			readParkingPointFromPlugin();
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/TargetPointsHelper.java]: No plugin calls here

File: OsmAnd/src/net/osmand/plus/TargetPointsHelper.java [hunk 15]
```diff
@@ -362,6 +416,10 @@ public TargetPoint getWorkPoint() {
 		return workPoint;
 	}
 
+	public TargetPoint getParkingPoint() {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/TargetPointsHelper.java]: No plugin call here

File: OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java [hunk 16]
```diff
@@ -682,6 +842,45 @@ protected void createDefaultCategories() {
 		addEmptyCategory(context.getString(R.string.shared_string_others));
 	}
 
+	public FavoriteGroup createPersonalGroup() {
+		FavoriteGroup group = new FavoriteGroup();
+		group.name = PERSONAL;
+		group.visible = true;
+		group.color = 0;
+		FavouritePoint fPoint;
+		TargetPointsHelper helper = context.getTargetPointsHelper();
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java]: TargetPointsHelper should not deal anymore with Home/Work! This code should be deleted

File: OsmAnd/src/net/osmand/plus/AppInitializer.java (hunk 4 above)
Comment (Dima-1) [reply OsmAnd/src/net/osmand/plus/AppInitializer.java]: It's for migration when Fix will be applied version number will be different

File: OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java (hunk 3 above)
Comment (Dima-1) [reply OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java]: It's hardcoded string HOME always first

Comment (max-klaus) [reply OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java]: On the UI we need to operate with translations and not with constants.

File: OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java (hunk 2 above)
Comment (Dima-1) [reply OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java]: Yes. I think the same

File: OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java (hunk 1 above)
Comment (Dima-1) [reply OsmAnd/src/net/osmand/plus/FavouritesDbHelper.java]: Not. It's left from first try.

**Pull request closed without merging**